src/V4Router.sol

```diff
@@ -0,0 +1,165 @@
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+
+import {PathKey, PathKeyLib} from "./libraries/PathKey.sol";
+import {CalldataDecoder} from "./libraries/CalldataDecoder.sol";
+import {IV4Router} from "./interfaces/IV4Router.sol";
+import {BaseActionsRouter} from "./base/BaseActionsRouter.sol";
+import {DeltaResolver} from "./base/DeltaResolver.sol";
+import {Actions} from "./libraries/Actions.sol";
+
+/// @title UniswapV4Router
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+/// @dev the entry point to executing actions in this contract is calling `BaseActionsRouter._executeActions`
+/// An inheriting contract should call _executeActions at the point that they wish actions to be executed
+abstract contract V4Router is IV4Router, BaseActionsRouter, DeltaResolver {
+    using PathKeyLib for PathKey;
+    using CalldataDecoder for bytes;
+    using TransientStateLibrary for IPoolManager;
+
+    constructor(IPoolManager _poolManager) BaseActionsRouter(_poolManager) {}
+
+    // TODO native support !!
+    function _handleAction(uint256 action, bytes calldata params) internal override {
+        // swap actions and payment actions in different blocks for gas efficiency
+        if (action < Actions.SETTLE) {
+            if (action == Actions.SWAP_EXACT_IN) {
+                _swapExactInput(abi.decode(params, (IV4Router.ExactInputParams)));
+            } else if (action == Actions.SWAP_EXACT_IN_SINGLE) {
+                _swapExactInputSingle(abi.decode(params, (IV4Router.ExactInputSingleParams)));
+            } else if (action == Actions.SWAP_EXACT_OUT) {
+                _swapExactOutput(abi.decode(params, (IV4Router.ExactOutputParams)));
+            } else if (action == Actions.SWAP_EXACT_OUT_SINGLE) {
+                _swapExactOutputSingle(abi.decode(params, (IV4Router.ExactOutputSingleParams)));
+            } else {
+                revert UnsupportedAction(action);
+            }
+        } else {
+            if (action == Actions.SETTLE_ALL) {
+                // equivalent: abi.decode(params, (Currency))
+                Currency currency;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                }
+
+                int256 delta = poolManager.currencyDelta(address(this), currency);
+                if (delta > 0) revert InvalidDeltaForAction();
+
+                // TODO support address(this) paying too
+                // TODO should it have a maxAmountOut added slippage protection?
+                _settle(currency, _msgSender(), uint256(-delta));
+            } else if (action == Actions.TAKE_ALL) {
+                // equivalent: abi.decode(params, (Currency, address))
+                Currency currency;
+                address recipient;
+                assembly ("memory-safe") {
```

Comment:
[#199](https://github.com/Uniswap/v4-periphery/issues/199) just tagging that 199 should also put these in a lib so we can use in both contracts